packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/mod.rs

```diff
@@ -271,6 +279,11 @@ impl<C: AsClient> EntityStore for PostgresStore<C> {
                 .change_context(UpdateError)?,
         );
 
+        let previous_entity = transaction
```

Comment:
Whereabouts do we think that a race condition would apppear? 

If we worry about another `update` coming in concurrently, the two transactions would be looking for the same row at the transaction snapshot, which may cause an old value to be read. We may have to consult https://www.postgresql.org/docs/current/sql-select.html#SQL-FOR-UPDATE-SHARE

Reply:
Yes, there could be a race condition. I think we should get the previous entity as part of the update SQL statement.

Reply:
Repro steps for data race:
Running the following two queries one after another and not committing/rolling back either of them:
```sql
BEGIN;
WITH entity AS(
	SELECT entity_id, entity_type_version_id, entities.version, type_ids.base_uri, type_ids.version, owned_by_id, properties
	FROM entities
	INNER JOIN type_ids ON type_ids.version_id = entities.entity_type_version_id
	WHERE entity_id = '7f873b40-12d1-45bc-a5eb-5531a877da2d' AND entities.version = (
		SELECT MAX("version")
		FROM entities
		WHERE entity_id = '7f873b40-12d1-45bc-a5eb-5531a877da2d'
	) FOR UPDATE
)
insert into entities (entity_id, "version", entity_type_version_id, properties, owned_by_id)
select entity_id, clock_timestamp(), entity_type_version_id, properties, owned_by_id from entity
returning *
```
and
```sql
BEGIN;
SELECT entity_id, entities.version, properties, type_ids.base_uri, type_ids.version, owned_by_id
FROM entities
INNER JOIN type_ids ON type_ids.version_id = entities.entity_type_version_id
WHERE entity_id = '7f873b40-12d1-45bc-a5eb-5531a877da2d' AND entities.version = (
	SELECT MAX("version")
	FROM entities
	WHERE entity_id = '7f873b40-12d1-45bc-a5eb-5531a877da2d'
) FOR UPDATE;
```
Where `7f873b40-12d1-45bc-a5eb-5531a877da2d` is just any entity ID in the database.
This shows that `FOR UPDATE` would not be adequate for locking the tables. I think for now, omitting any locking is the best cause of action until we figure out what to do. The transactions would likely run fast enough currently, so setting transaction serialization may be a good idea.

Reply:
> I think for now, omitting any locking is the best cause of action until we figure out what to do

Should I leave a todo here for now then, alongside a follow-up task?

Reply:
Are we not able to look up the previous entity as part of the SQL query?

I really do not like the idea of us either introducing locking, or knowingly leaving in a race condition which could break invariants. 

Like can we not express in SQL:
"use the owned_by_id and created_by_id of the most recent entity"?

Reply:
The problem is not being unable to express that, the problem lies in being able to fetch the most recent entity under transactions. Because of snapshotting, concurrent transactions would see the same "latest" entity and race for the last write (the new latest entity). Having a deleted table would likely help, as the `most recent entity` would be materialized in the `entities` table instead of having to aggregate, which means `FOR UPDATE` row locking would work as intended (sans any multi-shard citus issues)

Reply:
How do we solve this for incrementing versions of types?

Reply:
I've added a todo and a link to a follow-up task for now in https://github.com/hashintel/hash/pull/1197/commits/dc8d9246a233631ee3e8a75292668b485d4e388a